core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java

```diff
@@ -695,11 +690,16 @@ public boolean append(Object kbase, long koff, int klen, Object vbase, long voff
       assert (vlen % 8 == 0);
       assert (longArray != null);
 
-      if (numKeys == MAX_CAPACITY
-        // The map could be reused from last spill (because of no enough memory to grow),
-        // then we don't try to grow again if hit the `growthThreshold`.
-        || !canGrowArray && numKeys > growthThreshold) {
-        return false;
+      if (numKeys >= growthThreshold) {
+        if (longArray.size() / 2 == MAX_CAPACITY) {
```

Comment:
If the `currentKeysLen` above is the number of keys, it never equals to `currentSize / 2`. `currentSize / 2` is actually the `capacity` we want to allocate (but actually we allocate double of it for the array).

Once the number of keys reaches `growthThreshold` (i.e., `capacity * 0.5`), we go to grow the array or fail the append. So the number of keys is always less than or equal to `capacity * 0.5` which is `currentSize * 0.5 * 0.5`.